src/com/sun/jna/internal/Cleaner.java

```diff
@@ -44,65 +50,79 @@ public static Cleaner getCleaner() {
         return INSTANCE;
     }
 
+    // Guard for trackedObjects and cleanerThread. The readlock is utilized when
+    // the trackedObjects are manipulated, the writelock protectes starting and
+    // stopping the CleanerThread
+    private final ReadWriteLock cleanerThreadLock = new ReentrantReadWriteLock();
     private final ReferenceQueue<Object> referenceQueue;
-    private Thread cleanerThread;
-    private CleanerRef firstCleanable;
+    // Count of objects tracked by the cleaner. When >0 it means objects are
+    // being tracked by the cleaner and the cleanerThread must be running
+    private final AtomicLong trackedObjects = new AtomicLong();
+    // Map only serves as holder, so that the CleanerRefs stay hard referenced
+    // and quickly accessible for removal
+    @SuppressWarnings("MismatchedQueryAndUpdateOfCollection")
+    private final Map<CleanerRef,CleanerRef> map = new ConcurrentHashMap<>();
+    // Thread to handle the actual cleaning
+    private volatile Thread cleanerThread;
 
     private Cleaner() {
         referenceQueue = new ReferenceQueue<>();
     }
 
-    public synchronized Cleanable register(Object obj, Runnable cleanupTask) {
+    @SuppressWarnings("EmptySynchronizedStatement")
+    public Cleanable register(Object obj, Runnable cleanupTask) {
         // The important side effect is the PhantomReference, that is yielded
         // after the referent is GCed
-        return add(new CleanerRef(this, obj, referenceQueue, cleanupTask));
+        try {
+            return add(new CleanerRef(this, obj, referenceQueue, cleanupTask));
+        } finally {
+            synchronized (obj) {
+                // Used as a reachability fence for obj
+                // Ensure, that add completes before obj can be collected and
+                // the cleaner is run
+            }
+        }
     }
 
-    private synchronized CleanerRef add(CleanerRef ref) {
-        synchronized (referenceQueue) {
-            if (firstCleanable == null) {
-                firstCleanable = ref;
-            } else {
-                ref.setNext(firstCleanable);
-                firstCleanable.setPrevious(ref);
-                firstCleanable = ref;
-            }
-            if (cleanerThread == null) {
-                Logger.getLogger(Cleaner.class.getName()).log(Level.FINE, "Starting CleanerThread");
-                cleanerThread = new CleanerThread();
-                cleanerThread.start();
+    private CleanerRef add(CleanerRef ref) {
+        map.put(ref, ref);
+        cleanerThreadLock.readLock().lock();
+        try {
+            long count = trackedObjects.incrementAndGet();
+            if (cleanerThread == null && count > 0) {
```

Comment:
Because `register()` ensures that `ref` cannot be cleaned before `add()` has completed, we know that `count > 0` is always true here.